site/_posts/2020-08-07-locking-an-account-with-breached-password.md

```diff
@@ -0,0 +1,258 @@
+  error_log(var_export($response, TRUE));
+  return;
+}
+http_response_code(500);
+?>
+```
+
+Let's walk through this code, line by line.
+
+```php
+// ...
+require __DIR__. '/config.php';
+require __DIR__ . '/vendor/autoload.php';
+// ...
+```
+
+First, there are some required libraries and files.
+
+```php
+//...
+$headers = getallheaders();
+if (!$headers) {
+  error_log("Invalid authorization header.");
+  return;
+}
+
+$authorization_value = $headers['Authorization'];
+if ($authorization_value !== $authorization_header_value) {
+  error_log("Invalid authorization header value found: ".$authorization_value);
+  return;
+}
+//...
+```
+
+Then, the code checks the `Authorization` header. This ensures that only FusionAuth calls this webhook. For production, you would definitely want to use TLS as well.
+
+```php
+//...
+$input = file_get_contents('php://input');
+
+$obj = json_decode($input);
+//...
+```
+
+In these lines, the entire contents of the webhook payload are converted into a string. The string is then decoded into a JSON object for easier handling.
+
+```php
+//...
+if (!$obj) { 
+  error_log("Invalid JSON");
+  return;
+}
+
+type = $obj->event->type;
+if ($type !== "user.password.breach") {
+  error_log("Sorry, we only handle breached password events.");
+  return;
+}
+
+$user_id = $obj->event->user->id;
+if (!$user_id) {
+  error_log("No user id");
+  return;
+}
+//...
+```
+
+Next, validate the payload. If you don't get valid JSON, a password breach event, and a user id, simply log an error and return. Doing so allows the webhook and the login event to succeed. Check for the event type just in case there's a misconfiguration and our webhook is notified of other types of events.
+
+```php
+//...
+$client = new FusionAuth\FusionAuthClient($api_key, $fa_url);
+$response = $client->deactivateUser($user_id);
+if (!$response->wasSuccessful()) {
+  error_log("Response wasn't successful:");
+  error_log(var_export($response, TRUE));
+  return;
+}
+//...
+```
+
+Finally! This is where the action is. This code creates a new FusionAuth client. It then deactivates the user who logged in with a password found to be compromised. You could take other steps here, too. 
+
+You could do more within FusionAuth, by, for example:
+
+* Adding a date of deactivation to the `user.data` field
+* [Actioning the user](/docs/v1/tech/apis/actioning-users) for display in the administrative interface or to be queried via the API
+* Putting the user in a [group](/docs/v1/tech/core-concepts/groups) for future processing
+
+You could also integrate with other systems.
+
+* You could fire off an API call to another service that needs to know about this security violation.
+* The system could add an event to a streaming service, such as Kafka, for future analysis.
+* Your application could send an email to the user and their boss about the situation. Wouldn't be cool, but you could do it.
+
+```php
+//...
+http_response_code(500);
```

Comment:
As an aside, we do actually create an error event log when a webhook fails. So if you the webhook wanted to return JSON in a response body it would get logged in the event log for record. 

I may open an issue to better define expected return status codes. It would be kind of nice to allow the webhook to return a status code that says fail the request, but don't log an error. So maybe we could document that `500` is an error on your end, and a `4xx` is a fail, but only log it if debug is enabled for the webhook. 

https://github.com/FusionAuth/fusionauth-issues/issues/814